First chunk of the power9 integer enablement. Starting with new files,

munroesj52 opened the pull request:
dependant changed to vec_common_ppc.h, and more fixes. These will not
be activated until parts change Makefile.am.

2018-05-27 Steven Munroe <munroesj52@gmail.com>

        * src/vec_int16_ppc.h: New File.
        * src/vec_int32_ppc.h: New File.
        * src/vec_int64_ppc.h: New File.
        * src/vec_common_ppc.h (CONST_VINT128_H, CONST_VINT16_H,
        VEC_BYTE_L_DWH, VEC_BYTE_L_DWL): new Macro.
        * src/vec_f64_ppc.h: Minor fixup.

Signed-off-by: Steven Munroe <munroesj52@gmail.com>

ThinkOpenly commented on src/vec_int32_ppc.h:
```diff
@@ -0,0 +1,649 @@
+      "vpopcntw %0,%1;"
+      : "=v" (t)
+      : "v" (vra)
+      : );
```
`r` is not set here, but is returned later.  Oh, maybe that `t` is supposed to be `r`?

ThinkOpenly commented on src/vec_int32_ppc.h:
```diff
@@ -0,0 +1,649 @@
+}
+#else
+#undef vec_popcntw
+#define vec_popcntw __builtin_vec_vpopcntw
```
curious: what's going on here? might be worth a comment

ThinkOpenly commented on src/vec_int32_ppc.h:
```diff
@@ -0,0 +1,649 @@
+ *  propagated to each bit of each element.
+ *
+ *  @param vra a 128-bit vector treated as a vector signed int.
+ *  @param shb Shift amount in the range 0-31.
```
nit: do you want `Shift` capitalized? none of the others (line above and line below) have capitalization. (occurs elsewhere as well)

ThinkOpenly commented on src/vec_int32_ppc.h:
```diff
@@ -0,0 +1,649 @@
+    {
+      /* Load the shift const in a vector.  The element shifts require
+         a shift amount for each element. For the immediate form the
+         shift constant is splated to all elements of the
```
nit: s/splated/splatted/ ? (global)

ThinkOpenly commented on src/vec_int64_ppc.h:
```diff
@@ -0,0 +1,1128 @@
+ * added multiply word operations that produce the full doubleword
+ * product and full quadword add / subtract (with carry extend).
+ *
+ * Power9 (PowerISA 3.0B) adds the <B>Vector Multiply-Sum unsigned
```
s/unsigned/Unsigned/ ?

ThinkOpenly commented on src/vec_int64_ppc.h:
```diff
@@ -0,0 +1,1128 @@
+ *
+ * Power9 (PowerISA 3.0B) adds the <B>Vector Multiply-Sum unsigned
+ * Doubleword Modulo</B> instruction. This is not the expected
+ * multiple even/odd/modulo doubleword nor a full multiply modulo
```
s/multiple/multiply/

ThinkOpenly commented on src/vec_int64_ppc.h:
```diff
@@ -0,0 +1,1128 @@
+ * quadword. But with a few extra (permutes and splat zero)
+ * instructions you can get equivalent function.
+ *
+ * Most of these intrinsic (compiler built-ins) operations are defined
```
s/built-ins/built-in/

ThinkOpenly commented on src/vec_int64_ppc.h:
```diff
@@ -0,0 +1,1128 @@
+ *
+ * This header covers operations that are either:
+ *
+ * - Operations implemented in hardware instructions for later
```
s/Operations implemented/Implemented/

ThinkOpenly commented on src/vec_int64_ppc.h:
```diff
@@ -0,0 +1,1128 @@
+ * <altivec.n> provided by available compilers in common use.
+ * Examples include Count Leading Zeros, Population Count and Byte
+ * Reverse.
+ * - Are commonly used operations, not covered by the ABI or
```
s/Are commonly/Commonly/

ThinkOpenly commented on src/vec_int16_ppc.h:
```diff
@@ -0,0 +1,391 @@
+ * <altivec.n> provided by available compilers in common use.
+ * Examples include Count Leading Zeros, Population Count and Byte
+ * Reverse.
+ * - Are commonly used operations, not covered by the ABI or
```
s/Are commonly/Commonly/

ThinkOpenly commented on src/vec_int32_ppc.h:
```diff
@@ -0,0 +1,649 @@
+ * <altivec.n> provided by available compilers in common use.
+ * Examples include Count Leading Zeros, Population Count and Byte
+ * Reverse.
+ * - Are commonly used operations, not covered by the ABI or
```
s/Are commonly/Commonly/

ThinkOpenly commented on src/vec_int64_ppc.h:
```diff
@@ -0,0 +1,1128 @@
+ * currently implemented in <vec_int128_ppc.h> which resolves a
+ * dependency on Add Quadword. These functions (vec_msumudm,
+ * vec_muleud, vec_muloud) all produce a quadword results and need
+ * vec_adduqm to sum partial products and earlier power platforms.
```
s/and/on/ ?
s/power/POWER/

ThinkOpenly commented on src/vec_int64_ppc.h:
```diff
@@ -0,0 +1,1128 @@
+ *  elements of a and b are equal.
+ *
+ *  For Power8 (PowerISA 2.07B) or later use the vec_all_eq built-in predicate directly.
+ *  Otherwise case to unsigned word and use the same predicate generating (<B>vcmpequw</B>).
```
`case to` ?  maybe `cast to` ?

ThinkOpenly commented on src/vec_int64_ppc.h:
```diff
@@ -0,0 +1,1128 @@
+ *  elements of a > b.
+ *
+ *  For Power8 (PowerISA 2.07B) or later use the vec_all_eq built-in predicate directly.
+ *  Otherwise case to unsigned word and use the same predicate generating (<B>vcmpequw</B>).
```
s/case to/cast to/

ThinkOpenly commented on src/vec_int64_ppc.h:
```diff
@@ -0,0 +1,1128 @@
+ *  elements of a > b.
+ *
+ *  For Power8 (PowerISA 2.07B) or later use the vec_all_eq built-in predicate directly.
+ *  Otherwise case to unsigned word and use the same predicate generating (<B>vcmpequw</B>).
```
s/case to/cast to/

ThinkOpenly commented on src/vec_int64_ppc.h:
```diff
@@ -0,0 +1,1128 @@
+  result = vec_all_le(a, b);
+#else
+  vui32_t wt= { -1, -1, -1, -1};
+  vui64_t gt_bool= vec_cmpleud (a, b);
```
space before `=`, here and line above? (and globally)

ThinkOpenly commented on src/vec_int64_ppc.h:
```diff
@@ -0,0 +1,1128 @@
+      break;
+    case 3:
+      result = vec_xxpermdi (vra, vrb, 3);
+      break;
```
curious:  could the above case statements all be combined into
```
result = vec_xxpermdi (vra, vrb, ctl & 3);
```
?

ThinkOpenly commented on src/vec_int64_ppc.h:
```diff
@@ -0,0 +1,1128 @@
+#ifndef vec_vpopcntd
+  __asm__(
+      "vpopcntd %0,%1;"
+      : "=v" (t)
```
s/`t`/`r`/ ?

ThinkOpenly commented on src/vec_int64_ppc.h:
```diff
@@ -0,0 +1,1128 @@
+#endif
+
+
+/** \brief Vector Shift left Doubleword Immediate.
```
should `left` be capitalized? here and 2 lines below

ThinkOpenly commented on src/vec_int64_ppc.h:
```diff
@@ -0,0 +1,1128 @@
+       result[1] = vra[1];
+       result[0] = vra[1];
+       */
+      result = vec_permdi (vra, vra, 3);
```
could both case statements be combined into a single statement:
```
result = vec_permdi (vra, vra, (ctl & 1) * 3);
```

ThinkOpenly commented on src/vec_int64_ppc.h:
```diff
@@ -0,0 +1,1128 @@
+
+/** \brief Vector Shift Right Doubleword Immediate.
+ *
+ *	Vector Shift right Doublewords each element [0-1], 0-63 bits,
```
s/right/Right/

ThinkOpenly commented on src/vec_int64_ppc.h:
```diff
@@ -0,0 +1,1128 @@
+  /* Isolate the high dword so that bits from the low dword,
+   * do not contaminate the result.  */
+  vr_h = vec_andc ((vui8_t)vra, (vui8_t)sel_mask);
+#if 0
```
what's up here?

ThinkOpenly commented on src/vec_int64_ppc.h:
```diff
@@ -0,0 +1,1128 @@
+
+  /* constrain the dword shift amounts to 0-63.  */
+  vsh_l = vec_and ((vui8_t)vrb, (vui8_t)shft_mask);
+  /* Isolate the high dword so that bits from the low dword,
```
s/,//

munroesj52 commented:
On Tue, 2018-05-29 at 06:55 -0700, Paul Clarke wrote:
> @ThinkOpenly requested changes on this pull request.
> I'm not sure how nit-picky to be on formatting (whitespace,
> capitalization, punctuation).
> In src/vec_int16_ppc.h:
> > @@ -0,0 +1,391 @@
> +/*
> + Copyright [2018] IBM Corporation.
> Is there a blessed form for the copyright line? I usually see
> Copyright (c) IBM 2018.

This was consistent but different from sphde which includes the (c).
Added the (c) globally, but not sure moving the year is necessary.
Should look into that ant open an separate issue if required.

> In src/vec_int16_ppc.h:
> > + Contributors:
> +      IBM Corporation, Steven Munroe
> +      Created on: Apr 06, 2018
> + */
> +
> +#ifndef VEC_INT16_PPC_H_
> +#define VEC_INT16_PPC_H_
> +
> +#include <vec_char_ppc.h>
> +
> +/*!
> + * \file  vec_int16_ppc.h
> + * \brief Header package containing a collection of 128-bit SIMD
> + * operations over 16-bit integer elements.
> + *
> + * Most vector short (16-bit integer) operations are are already
> s/are are/are/
OK

> In src/vec_int16_ppc.h:
> > + * operations over 16-bit integer elements.
> + *
> + * Most vector short (16-bit integer) operations are are already
> + * covered by the original VMX (AKA Altivec) instructions.
> + * VMX intrinsic (compiler built-ins) operations are defined in
> + * <altivec.h> and described in the compiler documentation.
> + *
> + * \note The compiler disables associated <altivec.h> built-ins if
> the
> + * <B>mcpu</B> target does not enable the specific instruction.
> + * For example if you compile with <B>-mcpu=power7</B>, vec_vclz and
> + * vec_vclzh will not be defined.  But vec_clzh is always defined in
> + * this header, will generate the minimum code, appropriate for the
> + * target, and produce correct results.
> + *
> + * This header serves to fill in functional gaps for older
> + * (Power7, Power8) processors and provides a in-line assembler
> (nit) s/ a / an /
OK

> In src/vec_int16_ppc.h:
> > + *
> + * \note The compiler disables associated <altivec.h> built-ins if
> the
> + * <B>mcpu</B> target does not enable the specific instruction.
> + * For example if you compile with <B>-mcpu=power7</B>, vec_vclz and
> + * vec_vclzh will not be defined.  But vec_clzh is always defined in
> + * this header, will generate the minimum code, appropriate for the
> + * target, and produce correct results.
> + *
> + * This header serves to fill in functional gaps for older
> + * (Power7, Power8) processors and provides a in-line assembler
> + * implementation for older compilers that do not
> + * provide the build-ins.
> + *
> + * This header covers operations that are either:
> + *
> + * - Operations implemented in hardware instructions for later
> s/Operations implemented/Implemented/
> In src/vec_int16_ppc.h:
> > + *  elements.
> + *  @return 128-bit vector with the population count for each
> halfword
> + *  element.
> + */
> +#ifndef vec_popcnth
> +static inline vui16_t
> +vec_popcnth (vui16_t vra)
> +{
> +  vui16_t r;
> +#ifdef _ARCH_PWR8
> +#ifndef vec_vpopcnth
> +  __asm__(
> +      "vpopcnth %0,%1;"
> +      : "=v" (t)
> +      : "v" (vra)
> +      : );
> r is not set here, but is later returned. I think that t is supposed
> to be r.
Ya I changed from 't' to 'r' (result) to be consistent across the
platform implementations. The testing for this is a bitch as it only
shows up ofr specific GCC version and -mcpu= combinations.


> In src/vec_int32_ppc.h:
> > +#define VEC_INT32_PPC_H_
> +
> +#include <vec_int16_ppc.h>
> +
> +/*!
> + * \file  vec_int32_ppc.h
> + * \brief Header package containing a collection of 128-bit SIMD
> + * operations over 32-bit integer elements.
> + *
> + * Most vector int (32-bit integer word) operations are implemented
> + * with PowerISA VMX instructions either defined by the original VMX
> + * (AKA Altivec) or added to later versions of the PowerISA.
> + * Power8 added several multiply word operations not included in the
> + * original VMX.
> + *
> + * Also some useful word wise merge, shift, and splat operations
> where
> s/where/were/
OK

> In src/vec_int32_ppc.h:
> > + * original VMX.
> + *
> + * Also some useful word wise merge, shift, and splat operations
> where
> + * added with VSX in PowerISA 2.06B.
> + * Most of these intrinsic (compiler built-ins) operations are
> defined
> + * in <altivec.h> and described in the compiler documentation.
> + *
> + * \note The compiler disables associated <altivec.h> built-ins if
> the
> + * <B>mcpu</B> target does not enable the specific instruction.
> + * For example if you compile with <B>-mcpu=power7</B>, vec_vclz and
> + * vec_vclzw will not be defined.  But vec_clzw is always defined in
> + * this header, will generate the minimum code, appropriate for the
> + * target, and produce correct results.
> + *
> + * Most of these operations are implemented in a single instruction
> + * on newer (Power8/Power9) processors.
> technically, references to the architecture should be all caps,
> `POWER8/POWER9", here and just below. Globally, actually.
Ya Ya. Technically it is POWER if we are talking about an IBM PowerISA
implementation and Power if we are talking about any (including none
IBM) PowerISA implementations. So far POWER7/8/9 are IBM Server only,
but that could change.

OK

> In src/vec_int32_ppc.h:
> > + * <B>mcpu</B> target does not enable the specific instruction.
> + * For example if you compile with <B>-mcpu=power7</B>, vec_vclz and
> + * vec_vclzw will not be defined.  But vec_clzw is always defined in
> + * this header, will generate the minimum code, appropriate for the
> + * target, and produce correct results.
> + *
> + * Most of these operations are implemented in a single instruction
> + * on newer (Power8/Power9) processors.
> + * This header serves to fill in functional gaps for older
> + * (Power7, Power8) processors and provides a in-line assembler
> + * implementation for older compilers that do not
> + * provide the build-ins.
> + *
> + * This header covers operations that are either:
> + *
> + * - Operations implemented in hardware instructions for later
> s/Operations implemented/Implemented/
OK

> In src/vec_int32_ppc.h:
> > + * <altivec.n> provided by available compilers in common use.
> + * Examples include Count Leading Zeros, Population Count and Byte
> + * Reverse.
> + * - Are commonly used operations, not covered by the ABI or
> + * <altivec.h>, and require multiple instructions or
> + * are not obvious.
> + * Examples include the shift immediate operations.
> + *
> + */
> +
> +/** \brief Vector Count Leading Zeros word.
> + *
> + *  Count the number of leading '0' bits (0-32) within each word
> + *  element of a 128-bit vector.
> + *
> + *  For Power8 (PowerISA 2.07B) or later use the Vector Count
> Leading
> s/Power8/POWER8/
OK

> In src/vec_int32_ppc.h:
> > +static inline vui32_t
> +vec_clzw (vui32_t vra)
> +{
> +  vui32_t r;
> +#ifdef _ARCH_PWR8
> +#ifndef vec_vclzw
> +  __asm__(
> +      "vclzw %0,%1;"
> +      : "=v" (r)
> +      : "v" (vra)
> +      : );
> +#else
> +  r = vec_vclzw (vra);
> +#endif
> +#else
> +//#warning Implememention pre power8
> s/power8/POWER8/
OK

> In src/vec_int32_ppc.h:
> > +  vui32_t n, nt, y, x, s, m;
> +  vui32_t z= { 0,0,0,0};
> +  vui32_t one = { 1,1,1,1};
> +
> +  /* n = 32 s = 16 */
> +  s = vec_splat_u32(8);
> +  s = vec_add (s, s);
> +  n = vec_add (s, s);
> +
> +  x = vra;
> +  /* y=x>>16 if (y!=0) (n=n-16 x=y)  */
> +  y = vec_sr(x, s);
> +  nt = vec_sub(n,s);
> +  m = (vui32_t)vec_cmpgt(y, z);
> +  s = vec_sr(s,one);
> +  x = vec_sel (x , y, m);
> extra space after x. Also, sometimes you use a space before
> parentheses, sometimes not. Sometimes you have a space after a comma,
> sometimes not. Do you want to be globally consistent?
Fixed the leading ' ,' which was a pasto, but the GNU formatter does no
seem to care about leading ' (' or not. If you are really upset by this
open a separate issue ;)

> In src/vec_int32_ppc.h:
> > +{
> +  vui32_t r;
> +#ifdef _ARCH_PWR8
> +#ifndef vec_vclzw
> +  __asm__(
> +      "vclzw %0,%1;"
> +      : "=v" (r)
> +      : "v" (vra)
> +      : );
> +#else
> +  r = vec_vclzw (vra);
> +#endif
> +#else
> +//#warning Implememention pre power8
> +  vui32_t n, nt, y, x, s, m;
> +  vui32_t z= { 0,0,0,0};
> space after brace. here and next line.
The SDK GNU formatter does bazaar things to these initializers. Fixed
the ones you pointed out. 

Unless you want to fix the SDK I am not planning to do this globally.

> In src/vec_int32_ppc.h:
> > +      "vmulosw %0,%1,%2;\n"
> +#else
> +      "vmulesw %0,%1,%2;\n"
> +#endif
> +      : "=v" (res)
> +      : "v" (a),
> +      "v" (b)
> +      : );
> +#endif
> +  return (res);
> +}
> +
> +/** \brief Vector multiply odd signed words.
> + *
> + * Multiple the odd words of two vector signed int values and return
> + * the signed long product of the odd words..
> s/.././
OK

> In src/vec_int32_ppc.h:
> > + *  elements.
> + *  @return 128-bit vector with the population count for each word
> + *  element.
> + */
> +#ifndef vec_popcntw
> +static inline vui32_t
> +vec_popcntw (vui32_t vra)
> +{
> +  vui32_t r;
> +#ifdef _ARCH_PWR8
> +#ifndef vec_vpopcntw
> +  __asm__(
> +      "vpopcntw %0,%1;"
> +      : "=v" (t)
> +      : "v" (vra)
> +      : );
> r is not set here, but is returned later. Oh, maybe that t is
> supposed to be r?
Ya Ya. fixed.

> In src/vec_int32_ppc.h:
> > +      : );
> +#else
> +  r = vec_vpopcntw (vra);
> +#endif
> +#else
> +//#warning Implememention pre power8
> +  vui32_t z= { 0,0,0,0};
> +  vui8_t x;
> +  x = vec_popcntb ((vui8_t)vra);
> +  r = vec_vsum4ubs (x, z);
> +#endif
> +  return (r);
> +}
> +#else
> +#undef vec_popcntw
> +#define vec_popcntw __builtin_vec_vpopcntw
> curious: what's going on here? might be worth a comment
Compiler bug in GCC8. added.

/* Work around for GCC PR85830.  */

Where needed.  OK?

> In src/vec_int32_ppc.h:
> > +    }
> +
> +  return (vui32_t) result;
> +}
> +
> +/** \brief Vector Shift Right Algebraic Word Immediate.
> + *
> + *  Vector Shift Right Algebraic Words each element [0-3],
> + *  0-31 bits, as specified by an immediate value.
> + *  The shift amount is a const unsigned int in the range 0-31.
> + *  A shift count of 0 returns the original value of vra.
> + *  Shift counts greater then 31 bits return the sign bit
> + *  propagated to each bit of each element.
> + *
> + *  @param vra a 128-bit vector treated as a vector signed int.
> + *  @param shb Shift amount in the range 0-31.
> nit: do you want Shift capitalized? none of the others (line above
> and line below) have capitalization. (occurs elsewhere as well)
Ya Ya. fixed.

> In src/vec_int32_ppc.h:
> > + *
> + *  @param vra a 128-bit vector treated as a vector signed int.
> + *  @param shb Shift amount in the range 0-31.
> + *  @return 128-bit vector signed int, shifted right shb bits.
> + */
> +static inline vi32_t
> +vec_srawi (vi32_t vra, const unsigned int shb)
> +{
> +  vui32_t lshift;
> +  vi32_t result;
> +
> +  if (shb < 32)
> +    {
> +      /* Load the shift const in a vector.  The element shifts
> require
> +         a shift amount for each element. For the immediate form the
> +         shift constant is splated to all elements of the
> nit: s/splated/splatted/ ? (global)
Ya the spell checker is no help with this.

But as it is "splatter" I will go with "splatted"!

Fixed globally. OK?

> In src/vec_int64_ppc.h:
> > + * The original VMX (AKA Altivec) did not define any doubleword
> element
> + * (long long integer or double float) operations.
> + * The VSX facility (introduced with Power7) added vector double
> float
> + * but did not add any integer doubleword (64-bit)
> operations.  However
> + * it did add a useful doubleword permute immediate and word wise;
> + * merge, shift, and splat immediate operations.
> + * Otherwise vector long int (64-bit elements) operations
> + * have to be implemented using VMX word and halfword element
> integer
> + * operations for Power7.
> + *
> + * Power8 (PowerISA 2.07B) adds important doubleword integer (add,
> + * subtract, compare, shift, rotate, ...) VMX operations. Power8
> also
> + * added multiply word operations that produce the full doubleword
> + * product and full quadword add / subtract (with carry extend).
> + *
> + * Power9 (PowerISA 3.0B) adds the <B>Vector Multiply-Sum unsigned
> s/unsigned/Unsigned/ ?
OK

> In src/vec_int64_ppc.h:
> > + * The VSX facility (introduced with Power7) added vector double
> float
> + * but did not add any integer doubleword (64-bit)
> operations.  However
> + * it did add a useful doubleword permute immediate and word wise;
> + * merge, shift, and splat immediate operations.
> + * Otherwise vector long int (64-bit elements) operations
> + * have to be implemented using VMX word and halfword element
> integer
> + * operations for Power7.
> + *
> + * Power8 (PowerISA 2.07B) adds important doubleword integer (add,
> + * subtract, compare, shift, rotate, ...) VMX operations. Power8
> also
> + * added multiply word operations that produce the full doubleword
> + * product and full quadword add / subtract (with carry extend).
> + *
> + * Power9 (PowerISA 3.0B) adds the <B>Vector Multiply-Sum unsigned
> + * Doubleword Modulo</B> instruction. This is not the expected
> + * multiple even/odd/modulo doubleword nor a full multiply modulo
> s/multiple/multiply/
OK

> In src/vec_int64_ppc.h:
> > + * Otherwise vector long int (64-bit elements) operations
> + * have to be implemented using VMX word and halfword element
> integer
> + * operations for Power7.
> + *
> + * Power8 (PowerISA 2.07B) adds important doubleword integer (add,
> + * subtract, compare, shift, rotate, ...) VMX operations. Power8
> also
> + * added multiply word operations that produce the full doubleword
> + * product and full quadword add / subtract (with carry extend).
> + *
> + * Power9 (PowerISA 3.0B) adds the <B>Vector Multiply-Sum unsigned
> + * Doubleword Modulo</B> instruction. This is not the expected
> + * multiple even/odd/modulo doubleword nor a full multiply modulo
> + * quadword. But with a few extra (permutes and splat zero)
> + * instructions you can get equivalent function.
> + *
> + * Most of these intrinsic (compiler built-ins) operations are
> defined
> s/built-ins/built-in/
Ya Ya

> In src/vec_int64_ppc.h:
> > + * <B>mcpu</B> target does not enable the specific instruction.
> + * For example if you compile with <B>-mcpu=power7</B>, vec_vclz and
> + * vec_vclzd will not be defined.  But vec_clzd is always defined in
> + * this header, will generate the minimum code, appropriate for the
> + * target, and produce correct results.
> + *
> + * Most of these
> + * operations are implemented in a single instruction on newer
> + * (Power8/Power9) processors. So this header serves to fill in
> + * functional gaps for older (Power7, Power8) processors and
> provides
> + * a in-line assembler implementation for older compilers that do
> not
> + * provide the build-ins.
> + *
> + * This header covers operations that are either:
> + *
> + * - Operations implemented in hardware instructions for later
> s/Operations implemented/Implemented/
Ok

> In src/vec_int64_ppc.h:
> > + * (Power8/Power9) processors. So this header serves to fill in
> + * functional gaps for older (Power7, Power8) processors and
> provides
> + * a in-line assembler implementation for older compilers that do
> not
> + * provide the build-ins.
> + *
> + * This header covers operations that are either:
> + *
> + * - Operations implemented in hardware instructions for later
> + * processors and useful to programmers, on slightly older
> processors,
> + * even if the equivalent function requires more instructions.
> + * Examples include the multiply even/odd/modulo word operations.
> + * - Defined in the OpenPOWER ABI but <I>not</I> yet defined in
> + * <altivec.n> provided by available compilers in common use.
> + * Examples include Count Leading Zeros, Population Count and Byte
> + * Reverse.
> + * - Are commonly used operations, not covered by the ABI or
> s/Are commonly/Commonly/
OK

> In src/vec_int16_ppc.h:
> > + * (Power7, Power8) processors and provides a in-line assembler
> + * implementation for older compilers that do not
> + * provide the build-ins.
> + *
> + * This header covers operations that are either:
> + *
> + * - Operations implemented in hardware instructions for later
> + * processors and useful to programmers, on slightly older
> processors,
> + * even if the equivalent function requires more instructions.
> + * Examples include Count Leading Zeros, Population Count and Byte
> + * Reverse.
> + * - Defined in the OpenPOWER ABI but <I>not</I> yet defined in
> + * <altivec.n> provided by available compilers in common use.
> + * Examples include Count Leading Zeros, Population Count and Byte
> + * Reverse.
> + * - Are commonly used operations, not covered by the ABI or
> s/Are commonly/Commonly/
OK

> In src/vec_int32_ppc.h:
> > + * This header serves to fill in functional gaps for older
> + * (Power7, Power8) processors and provides a in-line assembler
> + * implementation for older compilers that do not
> + * provide the build-ins.
> + *
> + * This header covers operations that are either:
> + *
> + * - Operations implemented in hardware instructions for later
> + * processors and useful to programmers, on slightly older
> processors,
> + * even if the equivalent function requires more instructions.
> + * Examples include the multiply even/odd/modulo word operations.
> + * - Defined in the OpenPOWER ABI but <I>not</I> yet defined in
> + * <altivec.n> provided by available compilers in common use.
> + * Examples include Count Leading Zeros, Population Count and Byte
> + * Reverse.
> + * - Are commonly used operations, not covered by the ABI or
> s/Are commonly/Commonly/
OK

> In src/vec_int64_ppc.h:
> > + * processors and useful to programmers, on slightly older
> processors,
> + * even if the equivalent function requires more instructions.
> + * Examples include the multiply even/odd/modulo word operations.
> + * - Defined in the OpenPOWER ABI but <I>not</I> yet defined in
> + * <altivec.n> provided by available compilers in common use.
> + * Examples include Count Leading Zeros, Population Count and Byte
> + * Reverse.
> + * - Are commonly used operations, not covered by the ABI or
> + * <altivec.h>, and require multiple instructions or
> + * are not obvious.  Examples include the shift immediate
> operations.
> + *
> + * \note The Multiply even/odd doubleword operations are
> + * currently implemented in <vec_int128_ppc.h> which resolves a
> + * dependency on Add Quadword. These functions (vec_msumudm,
> + * vec_muleud, vec_muloud) all produce a quadword results and need
> + * vec_adduqm to sum partial products and earlier power platforms.
> s/and/on/ ?
> s/power/POWER/
OK

> In src/vec_int64_ppc.h:
> > +    {
> +       rr = vec_perm (r, r, permute);
> +       r= vec_and (r, rr);
> +    }
> +  result = (vui64_t)r;
> +#endif
> +  return (result);
> +}
> +
> +/** \brief Vector Compare all Equal Unsigned Doubleword.
> + *
> + *  Compare each unsigned long (64-bit) integer and return true if
> all
> + *  elements of a and b are equal.
> + *
> + *  For Power8 (PowerISA 2.07B) or later use the vec_all_eq built-in 
> predicate directly.
> + *  Otherwise case to unsigned word and use the same predicate
> generating (<B>vcmpequw</B>).
> case to ? maybe cast to ?
OK

> In src/vec_int64_ppc.h:
> > +  x = vec_or  (gt, y);
> +  /* Duplicate result word to dword width.  */
> +  y = vec_sld (c0, x, 12);
> +  r = vec_sel (x, y, c01);
> +  result = (vui64_t)r;
> +#endif
> +  return (result);
> +}
> +
> +/** \brief Vector Compare all Greater Than Unsigned Doubleword.
> + *
> + *  Compare each unsigned long (64-bit) integer and return true if
> all
> + *  elements of a > b.
> + *
> + *  For Power8 (PowerISA 2.07B) or later use the vec_all_eq built-in 
> predicate directly.
> + *  Otherwise case to unsigned word and use the same predicate
> generating (<B>vcmpequw</B>).
> s/case to/cast to/
OK

> In src/vec_int64_ppc.h:
> > +      : );
> +#endif
> +#else
> +  vui64_t r = vec_cmpgtud (a, b);
> +  result = vec_nor(r, r);;
> +#endif
> +  return (result);
> +}
> +
> +/** \brief Vector Compare all Less than equal Unsigned Doubleword.
> + *
> + *  Compare each unsigned long (64-bit) integer and return true if
> all
> + *  elements of a > b.
> + *
> + *  For Power8 (PowerISA 2.07B) or later use the vec_all_eq built-in 
> predicate directly.
> + *  Otherwise case to unsigned word and use the same predicate
> generating (<B>vcmpequw</B>).
> s/case to/cast to/
OK

> In src/vec_int64_ppc.h:
> > + *  @param b 128-bit vector treated as 2 x 64-bit unsigned long
> + *  integer (dword) elements.
> + *  @return boolean int for all 128-bits, true if all Greater Than,
> + *  false otherwise.
> + */
> +static inline
> +int
> +vec_cmpud_all_le (vui64_t a, vui64_t b)
> +{
> +  int result;
> +#ifdef _ARCH_PWR8
> +#if __GNUC__ >= 7
> +  result = vec_all_le(a, b);
> +#else
> +  vui32_t wt= { -1, -1, -1, -1};
> +  vui64_t gt_bool= vec_cmpleud (a, b);
> space before =, here and line above? (and globally)
Ya Ya

> In src/vec_int64_ppc.h:
> > +#ifdef _ARCH_PWR7
> +  switch (ctl & 3)
> +    {
> +#if __BYTE_ORDER__ == __ORDER_BIG_ENDIAN__
> +    case 0:
> +      result = vec_xxpermdi (vra, vrb, 0);
> +      break;
> +    case 1:
> +      result = vec_xxpermdi (vra, vrb, 1);
> +      break;
> +    case 2:
> +      result = vec_xxpermdi (vra, vrb, 2);
> +      break;
> +    case 3:
> +      result = vec_xxpermdi (vra, vrb, 3);
> +      break;
> curious: could the above case statements all be combined into
> result = vec_xxpermdi (vra, vrb, ctl & 3);
No, This is very deliberate and not changing.

> ?
> In src/vec_int64_ppc.h:
> > + *
> + *  @param vra 128-bit vector treated as 2 x 64-bit integer (dwords)
> + *  elements.
> + *  @return 128-bit vector with the population count for each dword
> + *  element.
> + */
> +#ifndef vec_popcntd
> +static inline vui64_t
> +vec_popcntd (vui64_t vra)
> +{
> +  vui64_t r;
> +#ifdef _ARCH_PWR8
> +#ifndef vec_vpopcntd
> +  __asm__(
> +      "vpopcntd %0,%1;"
> +      : "=v" (t)
> s/t/r/ ?
Ya Ya

> In src/vec_int64_ppc.h:
> > +      CONST_VINT64_DW(0x08090A0B0C0D0E0FUL, 0x0001020304050607UL);
> +#endif
> +  result = (vui64_t) vec_perm ((vui8_t) vra, (vui8_t) vra, (vui8_t)
> vconstp);
> +#endif
> +
> +  return (result);
> +}
> +
> +#ifndef vec_vsld
> +static inline vui64_t vec_vsld (vui64_t vra, vui64_t vrb);
> +static inline vui64_t vec_vsrd (vui64_t vra, vui64_t vrb);
> +static inline vi64_t vec_vsrad (vi64_t vra, vui64_t vrb);
> +#endif
> +
> +
> +/** \brief Vector Shift left Doubleword Immediate.
> should left be capitalized? here and 2 lines below
OK

> In src/vec_int64_ppc.h:
> > +     that.  */
> +  switch (ctl & 1)
> +    {
> +    case 0:
> +      /*
> +       result[1] = vra[0];
> +       result[0] = vra[0];
> +       */
> +      result = vec_permdi (vra, vra, 0);
> +      break;
> +    case 1:
> +      /*
> +       result[1] = vra[1];
> +       result[0] = vra[1];
> +       */
> +      result = vec_permdi (vra, vra, 3);
> could both case statements be combined into a single statement:
No. I want this expanded, in a straight forward way, that any one can
understand at a glance.

With the endian crap it becomes hard to understand otherwise.

> result = vec_permdi (vra, vra, (ctl & 1) * 3);
> In src/vec_int64_ppc.h:
> > +      break;
> +    case 1:
> +      /*
> +       result[1] = vra[1];
> +       result[0] = vra[1];
> +       */
> +      result = vec_permdi (vra, vra, 3);
> +      break;
> +    }
> +
> +  return (result);
> +}
> +
> +/** \brief Vector Shift Right Doubleword Immediate.
> + *
> + *	Vector Shift right Doublewords each element [0-1], 0-63
> bits,
> s/right/Right/
OK

> In src/vec_int64_ppc.h:
> > +      : "=v" (result)
> +      : "v" (vra),
> +      "v" (vrb)
> +      : );
> +#else
> +  vui8_t  vsh_h, vsh_l;
> +  vui8_t  vr_h, vr_l;
> +  vui64_t sel_mask = CONST_VINT128_DW(0, -1LL);
> +  vui64_t shft_mask = CONST_VINT128_DW(63, 63);
> +
> +  /* constrain the dword shift amounts to 0-63.  */
> +  vsh_l = vec_and ((vui8_t)vrb, (vui8_t)shft_mask);
> +  /* Isolate the high dword so that bits from the low dword,
> +   * do not contaminate the result.  */
> +  vr_h = vec_andc ((vui8_t)vra, (vui8_t)sel_mask);
> +#if 0
> what's up here?
Eliminated vec_and and added a comment that explained why it was
unnecessary.

> In src/vec_int64_ppc.h:
> > +#ifdef _ARCH_PWR8
> +  __asm__(
> +      "vsld %0,%1,%2;"
> +      : "=v" (result)
> +      : "v" (vra),
> +      "v" (vrb)
> +      : );
> +#else
> +  vui8_t  vsh_h, vsh_l;
> +  vui8_t  vr_h, vr_l;
> +  vui64_t sel_mask = CONST_VINT128_DW(0, -1LL);
> +  vui64_t shft_mask = CONST_VINT128_DW(63, 63);
> +
> +  /* constrain the dword shift amounts to 0-63.  */
> +  vsh_l = vec_and ((vui8_t)vrb, (vui8_t)shft_mask);
> +  /* Isolate the high dword so that bits from the low dword,
> s/,//
OK

> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub, or mute the thread.

munroesj52 commented:
Ok, took 3 tries to get all the nits cleaned up. Can be pull / merge this branch and move on or do I have start over with the fresh gork?

ThinkOpenly commented on src/vec_int16_ppc.h:
```diff
@@ -0,0 +1,392 @@
+ See the License for the specific language governing permissions and
+ limitations under the License.
+
+ vec_int32_ppc.h
```
should be vec_int16_ppc.h

ThinkOpenly commented on src/vec_int16_ppc.h:
```diff
@@ -0,0 +1,392 @@
+ vec_int32_ppc.h
+
+ Contributors:
+      IBM Corporation, Steven Munroe
```
aren't you a civilian now? :-)

ThinkOpenly commented on src/vec_int16_ppc.h:
```diff
@@ -0,0 +1,392 @@
+ *
+ * Most vector short (16-bit integer) operations are already
+ * covered by the original VMX (AKA Altivec) instructions.
+ * VMX intrinsic (compiler built-ins) operations are defined in
```
s/built-ins/built-in/

ThinkOpenly commented on src/vec_int16_ppc.h:
```diff
@@ -0,0 +1,392 @@
+ * <altivec.h> and described in the compiler documentation.
+ *
+ * \note The compiler disables associated <altivec.h> built-ins if the
+ * <B>mcpu</B> target does not enable the specific instruction.
```
did you want `-mcpu` here?  (missing the `-`, if so)

ThinkOpenly commented on src/vec_int16_ppc.h:
```diff
@@ -0,0 +1,392 @@
+ * This header serves to fill in functional gaps for older
+ * (POWER7, POWER8) processors and provides an in-line assembler
+ * implementation for older compilers that do not
+ * provide the build-ins.
```
s/build-ins/built-ins/

ThinkOpenly commented on src/vec_int16_ppc.h:
```diff
@@ -0,0 +1,392 @@
+ * Examples include Count Leading Zeros, Population Count and Byte
+ * Reverse.
+ * - Defined in the OpenPOWER ABI but <I>not</I> yet defined in
+ * <altivec.n> provided by available compilers in common use.
```
s/altivec.n/altivec.h/

ThinkOpenly commented on src/vec_int16_ppc.h:
```diff
@@ -0,0 +1,392 @@
+ * <altivec.h>, and require multiple instructions or
+ * are not obvious.
+ * Examples include the shift immediate operations.
+ *
```
nit: extra emtpy comment line

ThinkOpenly commented on src/vec_int16_ppc.h:
```diff
@@ -0,0 +1,392 @@
+ */
+
+
+/** \brief Count leading zeros for a vector unsigned short (halfword)
```
did you want lead-caps, as in `Count Leading Zeros` ?

ThinkOpenly commented on src/vec_int16_ppc.h:
```diff
@@ -0,0 +1,392 @@
+  r = vec_sel (nt , n, m);
+#endif
+
+  return (r);
```
nit: don't need the parens here.  Also in many other places.  not sure how much to care.

ThinkOpenly commented on src/vec_int16_ppc.h:
```diff
@@ -0,0 +1,392 @@
+  return (result);
+}
+
+/** \brief Vector Shift left Halfword Immediate.
```
s/left/Left/

ThinkOpenly commented on src/vec_int16_ppc.h:
```diff
@@ -0,0 +1,392 @@
+
+/** \brief Vector Shift left Halfword Immediate.
+ *
+ *	Vector Shift left Halfwords each element [0-7], 0-15 bits,
```
s/Vector Shift left Halfwords/Vector shift left halfwords/ ? (seems consistent with other places)

ThinkOpenly commented on src/vec_int16_ppc.h:
```diff
@@ -0,0 +1,392 @@
+
+/** \brief Vector Shift Right Halfword Immediate.
+ *
+ *	Vector Shift right Halfwords each element [0-7], 0-15 bits,
```
s/Vector Shift right Halfwords/Vector shift right halfwords/ ?

ThinkOpenly commented on src/vec_int16_ppc.h:
```diff
@@ -0,0 +1,392 @@
+
+/** \brief Vector Shift Right Algebraic Halfword Immediate.
+ *
+ *  Vector Shift Right Algebraic Halfwords each element [0-7],
```
capitalization again. I know it's getting old, just looking for consistency. If you don't care, we can let it go.

ThinkOpenly commented on src/vec_int16_ppc.h:
```diff
@@ -0,0 +1,392 @@
+  return (vi16_t) result;
+}
+
+
```
extra blank line

ThinkOpenly commented on src/vec_int16_ppc.h:
```diff
@@ -0,0 +1,392 @@
+    }
+  else
+    { /* shifts greater then 15 bits returns the sign bit propagated to
+         all bits.   This is equivalent to shift Right Algebraic of
```
real nit: extra space after the `.`

ThinkOpenly commented on src/vec_common_ppc.h:
```diff
@@ -1,5 +1,5 @@
 /*
- Copyright [2017] IBM Corporation.
+ Copyright (c) [2017] IBM Corporation.
```
this should be 2018, right?

ThinkOpenly commented on src/vec_f64_ppc.h:
```diff
@@ -1,5 +1,5 @@
 /*
- Copyright [2017] IBM Corporation.
+ Copyright (c) [2017] IBM Corporation.
```
this should be 2018, right?

munroesj52 commented:
On Mon, 2018-06-04 at 09:28 -0700, Paul Clarke wrote:
> @ThinkOpenly requested changes on this pull request.
> at least a couple worth addressing
> In src/vec_int16_ppc.h:
> > +/*
> + Copyright (c) [2018] IBM Corporation.
> +
> + Licensed under the Apache License, Version 2.0 (the "License");
> + you may not use this file except in compliance with the License.
> + You may obtain a copy of the License at
> +
> +    http://www.apache.org/licenses/LICENSE-2.0
> +
> + Unless required by applicable law or agreed to in writing, software
> + distributed under the License is distributed on an "AS IS" BASIS,
> + WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or
> implied.
> + See the License for the specific language governing permissions and
> + limitations under the License.
> +
> + vec_int32_ppc.h
> should be vec_int16_ppc.h
Ya Ya.

> In src/vec_int16_ppc.h:
> > + Licensed under the Apache License, Version 2.0 (the "License");
> + you may not use this file except in compliance with the License.
> + You may obtain a copy of the License at
> +
> +    http://www.apache.org/licenses/LICENSE-2.0
> +
> + Unless required by applicable law or agreed to in writing, software
> + distributed under the License is distributed on an "AS IS" BASIS,
> + WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or
> implied.
> + See the License for the specific language governing permissions and
> + limitations under the License.
> +
> + vec_int32_ppc.h
> +
> + Contributors:
> +      IBM Corporation, Steven Munroe
> aren't you a civilian now? :-)
Yes but, a lot of this code was written for arith128 (before I retired)
and copied from other files.

So ...

IBM Corporation, Steven Munroe	Initial Contribution.
Steven Munroe, additional functions.

? We can change this in a later commit once we decide.

> In src/vec_int16_ppc.h:
> > +      Created on: Apr 06, 2018
> + */
> +
> +#ifndef VEC_INT16_PPC_H_
> +#define VEC_INT16_PPC_H_
> +
> +#include <vec_char_ppc.h>
> +
> +/*!
> + * \file  vec_int16_ppc.h
> + * \brief Header package containing a collection of 128-bit SIMD
> + * operations over 16-bit integer elements.
> + *
> + * Most vector short (16-bit integer) operations are already
> + * covered by the original VMX (AKA Altivec) instructions.
> + * VMX intrinsic (compiler built-ins) operations are defined in
> s/built-ins/built-in/
> In src/vec_int16_ppc.h:
> > +#define VEC_INT16_PPC_H_
> +
> +#include <vec_char_ppc.h>
> +
> +/*!
> + * \file  vec_int16_ppc.h
> + * \brief Header package containing a collection of 128-bit SIMD
> + * operations over 16-bit integer elements.
> + *
> + * Most vector short (16-bit integer) operations are already
> + * covered by the original VMX (AKA Altivec) instructions.
> + * VMX intrinsic (compiler built-ins) operations are defined in
> + * <altivec.h> and described in the compiler documentation.
> + *
> + * \note The compiler disables associated <altivec.h> built-ins if
> the
> + * <B>mcpu</B> target does not enable the specific instruction.
> did you want -mcpu here? (missing the -, if so)
Could be either way. This is not ambiguous with/without the '-'

> In src/vec_int16_ppc.h:
> > + * Most vector short (16-bit integer) operations are already
> + * covered by the original VMX (AKA Altivec) instructions.
> + * VMX intrinsic (compiler built-ins) operations are defined in
> + * <altivec.h> and described in the compiler documentation.
> + *
> + * \note The compiler disables associated <altivec.h> built-ins if
> the
> + * <B>mcpu</B> target does not enable the specific instruction.
> + * For example if you compile with <B>-mcpu=power7</B>, vec_vclz and
> + * vec_vclzh will not be defined.  But vec_clzh is always defined in
> + * this header, will generate the minimum code, appropriate for the
> + * target, and produce correct results.
> + *
> + * This header serves to fill in functional gaps for older
> + * (POWER7, POWER8) processors and provides an in-line assembler
> + * implementation for older compilers that do not
> + * provide the build-ins.
> s/build-ins/built-ins/
OK

> In src/vec_int16_ppc.h:
> > + * target, and produce correct results.
> + *
> + * This header serves to fill in functional gaps for older
> + * (POWER7, POWER8) processors and provides an in-line assembler
> + * implementation for older compilers that do not
> + * provide the build-ins.
> + *
> + * This header covers operations that are either:
> + *
> + * - Implemented in hardware instructions for later
> + * processors and useful to programmers, on slightly older
> processors,
> + * even if the equivalent function requires more instructions.
> + * Examples include Count Leading Zeros, Population Count and Byte
> + * Reverse.
> + * - Defined in the OpenPOWER ABI but <I>not</I> yet defined in
> + * <altivec.n> provided by available compilers in common use.
> s/altivec.n/altivec.h/
OK

> In src/vec_int16_ppc.h:
> > + * This header covers operations that are either:
> + *
> + * - Implemented in hardware instructions for later
> + * processors and useful to programmers, on slightly older
> processors,
> + * even if the equivalent function requires more instructions.
> + * Examples include Count Leading Zeros, Population Count and Byte
> + * Reverse.
> + * - Defined in the OpenPOWER ABI but <I>not</I> yet defined in
> + * <altivec.n> provided by available compilers in common use.
> + * Examples include Count Leading Zeros, Population Count and Byte
> + * Reverse.
> + * - Commonly used operations, not covered by the ABI or
> + * <altivec.h>, and require multiple instructions or
> + * are not obvious.
> + * Examples include the shift immediate operations.
> + *
> nit: extra emtpy comment line
Ya Ya not a big issue

> In src/vec_int16_ppc.h:
> > + * even if the equivalent function requires more instructions.
> + * Examples include Count Leading Zeros, Population Count and Byte
> + * Reverse.
> + * - Defined in the OpenPOWER ABI but <I>not</I> yet defined in
> + * <altivec.n> provided by available compilers in common use.
> + * Examples include Count Leading Zeros, Population Count and Byte
> + * Reverse.
> + * - Commonly used operations, not covered by the ABI or
> + * <altivec.h>, and require multiple instructions or
> + * are not obvious.
> + * Examples include the shift immediate operations.
> + *
> + */
> +
> +
> +/** \brief Count leading zeros for a vector unsigned short
> (halfword)
> did you want lead-caps, as in Count Leading Zeros ?
I try to use caps then spelling out something might be a instruction.

But again, not critical.

> In src/vec_int16_ppc.h:
> > +  x = vec_sel (x , y, m);
> +  n = vec_sel (n , nt, m);
> +
> +  /* y=x>>1 if (y!=0) return (n=n-2)   */
> +  y = vec_sr(x, s);
> +  nt = vec_sub(n,s);
> +  nt = vec_sub(nt,s);
> +  m = (vui16_t)vec_cmpgt(y, z);
> +  n = vec_sel (n , nt, m);
> +
> +  /* else return (x-n)  */
> +  nt = vec_sub (n, x);
> +  r = vec_sel (nt , n, m);
> +#endif
> +
> +  return (r);
> nit: don't need the parens here. Also in many other places. not sure
> how much to care.

Something in the compiler was yelling at me, when I did not.

Not changing this now.

> In src/vec_int16_ppc.h:
> > +  result = vec_revb (vra);
> +#endif
> +#else
> +#if __BYTE_ORDER__ == __ORDER_BIG_ENDIAN__
> +  const vui64_t vconstp = CONST_VINT64_DW(0x0100030205040706UL,
> 0x09080B0A0D0C0F0EUL);
> +#else
> +  const vui64_t vconstp =
> +      CONST_VINT64_DW(0x0E0F0C0D0A0B0809UL, 0x0607040502030001UL);
> +#endif
> +  result = (vui16_t) vec_perm ((vui8_t) vra, (vui8_t) vra, (vui8_t)
> vconstp);
> +#endif
> +
> +  return (result);
> +}
> +
> +/** \brief Vector Shift left Halfword Immediate.
> s/left/Left/
> In src/vec_int16_ppc.h:
> > +#else
> +#if __BYTE_ORDER__ == __ORDER_BIG_ENDIAN__
> +  const vui64_t vconstp = CONST_VINT64_DW(0x0100030205040706UL,
> 0x09080B0A0D0C0F0EUL);
> +#else
> +  const vui64_t vconstp =
> +      CONST_VINT64_DW(0x0E0F0C0D0A0B0809UL, 0x0607040502030001UL);
> +#endif
> +  result = (vui16_t) vec_perm ((vui8_t) vra, (vui8_t) vra, (vui8_t)
> vconstp);
> +#endif
> +
> +  return (result);
> +}
> +
> +/** \brief Vector Shift left Halfword Immediate.
> + *
> + *	Vector Shift left Halfwords each element [0-7], 0-15 bits,
> s/Vector Shift left Halfwords/Vector shift left halfwords/ ? (seems
> consistent with other places)
> In src/vec_int16_ppc.h:
> > +
> +      /* Vector Shift right bytes based on the lower 4-bits of
> +         corresponding element of lshift.  */
> +      result = vec_vslh (vra, lshift);
> +    }
> +  else
> +    { /* shifts greater then 15 bits return zeros.  */
> +      result = vec_xor ((vui16_t) vra, (vui16_t) vra);
> +    }
> +
> +  return (vui16_t) result;
> +}
> +
> +/** \brief Vector Shift Right Halfword Immediate.
> + *
> + *	Vector Shift right Halfwords each element [0-7], 0-15
> bits,
> s/Vector Shift right Halfwords/Vector shift right halfwords/ ?
Vector Shift Right Halfword is an instruction in the PowerISA.

Vector Shift Right Halfword Immediate should be an instruction.

I can change to comment to:

 * Shift right each halfword element [0-7], 0-15 bits,

is that clearer.

> In src/vec_int16_ppc.h:
> > +	lshift = vec_splats ((unsigned short) shb);
> +
> +      /* Vector Shift right bytes based on the lower 4-bits of
> +         corresponding element of lshift.  */
> +      result = vec_vsrh (vra, lshift);
> +    }
> +  else
> +    { /* shifts greater then 15 bits return zeros.  */
> +      result = vec_xor ((vui16_t) vra, (vui16_t) vra);
> +    }
> +  return (vui16_t) result;
> +}
> +
> +/** \brief Vector Shift Right Algebraic Halfword Immediate.
> + *
> + *  Vector Shift Right Algebraic Halfwords each element [0-7],
> capitalization again. I know it's getting old, just looking for
> consistency. If you don't care, we can let it go.

First lets make it work, correctly. then we can make it perfect.

If you want to make another pass for documentation grammar and spell
you can open an separate issue.

> In src/vec_int16_ppc.h:
> > +      /* Vector Shift Right Algebraic Halfwords based on the lower
> 4-bits
> +         of corresponding element of lshift.  */
> +      result = vec_vsrah (vra, lshift);
> +    }
> +  else
> +    { /* shifts greater then 15 bits returns the sign bit propagated
> to
> +         all bits.   This is equivalent to shift Right Algebraic of
> +         15 bits.  */
> +      lshift = (vui16_t) vec_splat_s16(15);
> +      result = vec_vsrah (vra, lshift);
> +    }
> +
> +  return (vi16_t) result;
> +}
> +
Ya Ya
> +
> extra blank line
> In src/vec_int16_ppc.h:
> > +      /* Load the shift const in a vector.  The element shifts
> require
> +         a shift amount for each element. For the immediate form the
> +         shift constant is splatted to all elements of the
> +         shift control.  */
> +      if (__builtin_constant_p(shb))
> +	lshift = (vui16_t) vec_splat_s16(shb);
> +      else
> +	lshift = vec_splats ((unsigned short) shb);
> +
> +      /* Vector Shift Right Algebraic Halfwords based on the lower
> 4-bits
> +         of corresponding element of lshift.  */
> +      result = vec_vsrah (vra, lshift);
> +    }
> +  else
> +    { /* shifts greater then 15 bits returns the sign bit propagated
> to
> +         all bits.   This is equivalent to shift Right Algebraic of
> real nit: extra space after the .
Ya Ya 
Drepper beat intoi me "exactly 2 spaces after the '.'.

I still have some scars

In src/vec_common_ppc.h:
> > @@ -1,5 +1,5 @@
>  /*
> - Copyright [2017] IBM Corporation.
> + Copyright (c) [2017] IBM Corporation.
> this should be 2018, right?
> In src/vec_f64_ppc.h:
> > @@ -1,5 +1,5 @@
>  /*
> - Copyright [2017] IBM Corporation.
> + Copyright (c) [2017] IBM Corporation.
> this should be 2018, right?
most of this file existed before. If you insist:

 Copyright (c) [2017, 2018] IBM Corporation.

There is a lot more backed up behind this. It is time to move on.

> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub, or mute the thread.

munroesj52 commented:
moving on to next set of changes
